Support spackbot rebuilding all specs from source

scottwittenburg opened the pull request:
Support in spack's gitlab pipelines allowing maintainers to ask spackbot to rebuild everything from source.  This should help test core changes to spack that don't otherwise change the hashes of tested specs.

- [x] Allow overriding --prune-dag cli opt with env var
- [x] Use job variable to optionally prevent rebuild jobs early exit behavior
- [x] Supply new install argument to insist deps are always installed from binary, but package is only installed from source.

Steps to get `@spackbot rebuild everything`:

1. work out installation from concrete spec.json issue with #32537 and merge it
2. merge this PR
3. merge spackbot PR [#67](https://github.com/spack/spackbot/pull/67)
4. merge spack-intrastructure PR [#319](https://github.com/spack/spack-infrastructure/pull/319)

scottwittenburg committed:
Support spackbot rebuilding all specs from source

- [x] Allow overriding --prune-dag cli opt with env var
- [x] Use job variable to optionally prevent rebuild jobs early
exit behavior
- [ ] Supply new install argument to insist deps are always
installed from binary, but package is only installed from source.

scottwittenburg committed:
gitlab: use new buildcache install option

scottwittenburg committed:
gitlab: fix bug w/ untouched pruning

scottwittenburg committed:
fix style

alalazo commented:
My understanding from last meeting is that #32537 is not needed to have pipelines rebuild all specs. It's an orthogonal feature that ensures each job in the Gitlab pipeline will build from sources only the root node and not its dependencies. Am I correct saying that, as such, that PR can be merged independently (i.e. before or after "rebuild all") ?

alalazo commented:
> No, we do need that first, unless we change our approach to "spackbot rebuild everything".

What is in that PR that is needed for rebuilding everything?

scottwittenburg commented:
> > No, we do need that first, unless we change our approach to "spackbot rebuild everything".
> 
> What is in that PR that is needed for rebuilding everything?

A single `spack install` command that insists on rebuilding the package from source (even if a binary mirror has the spec with the same hash), while the dependencies are still installed from one of the binary mirrors.

scottwittenburg commented:
fyi @zackgalbreath

alalazo commented:
> A single spack install command 

That's the part I can't currently figure out from code. Why do we need a single command?

scottwittenburg commented:
fyi @kwryankrattiger

scottwittenburg committed:
ci: install from hash rather than json file

scottwittenburg committed:
gitlab ci: Mirror fix when rebuilding everything

When doing a "rebuild everything" pipeline, make sure that each
install job only consumes binary dependencies from the mirror
being populated by the current pipeline.  This avoids using, e.g.
binaries from develop, when rebuilding everything on a PR.

scottwittenburg committed:
gitlab ci: fix how broken specs affect rebuild everything

When running a pipeline to rebuild everything, do not die because
we generated a hash on the broken specs list.  Instead only warn in
that case.

scottwittenburg commented:
@alalazo Should I retry the failed `ci / unit-tests / ubuntu (2.7, original, false)` and `ci / unit-tests / macos / (3.8)` checks?  I don't see any tests actually marked as failures in the output.  Maybe I'm just missing it though.

alalazo commented:
The test failing is:
```
lib/spack/spack/test/llnl/util/lock.py::test_write_lock_timeout_on_write_ranges_2[/run/user/1001]
```
which might be spurious. I'll trigger a re-run.

scottwittenburg commented:
Seems like the ubuntu python 2.7 and macos python 3.8 tests timed out this time.  I'll retry again.

scottwittenburg commented:
Retrying the ubunty python 2.7 check passed on the most recent retry.  However, the macos python 3.8 check is still failing, but I still don't see the error in there.

haampie commented on lib/spack/spack/cmd/ci.py:
```diff
@@ -526,10 +547,8 @@ def ci_rebuild(args):
-    # json file.
-    install_args.extend(["-f", job_spec_json_path])
+    # Identify spec to install by hash
+    install_args.append("/{0}".format(job_spec.dag_hash()))
```
how does this work with "install missing compilers"? then `job_spec` is neither in the db nor in the environment so it fails? or not?

scottwittenburg replied:
Unfortunately compiler bootstrapping pipelines haven't worked ever since we started using the `spack.lock` to communicate what to build to downstream jobs.

haampie commented on lib/spack/spack/cmd/ci.py:
```diff
@@ -526,10 +547,8 @@ def ci_rebuild(args):
     if compiler_action != "FIND_ANY":
         install_args.append("--no-add")
```
also, what's the point here actually?

haampie replied:
seems like the comment above is wrong?

scottwittenburg replied:
I don't recall why we wouldn't want `--no-add` in all cases.  Since we never encounter this in practice any more, we just end up always adding this flag.  We use the flag because when the purpose of the current job is to install a dependency of one of the roots of the environment, we just want to install it, not add it as a root itself.

scottwittenburg commented:
Retrying the unit-tests / macos (3.8) as it failed mysteriously again.

scottwittenburg committed:
Replace broken no-args tty.die() with sys.exit(1)

scottwittenburg commented:
Python 2.7 timed out again, retrying.
